### PR TITLE
blog: portforwarding is awesome

### PR DESCRIPTION
**Pull request by nicks**
Hello @ellenkorbes, @milas,

Please review the following commits I made in branch nicks/portforward:

226c25c826f14fec809ebbbb1ef9c25e9d748847 (2021-07-07 14:55:06 -0400)
blog: portforwarding is awesome

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  - networking
+---
+
+In the Unix mindset, every program reads and writes files. So it's important to have a toolkit of basic
+tools like `cat` for working with files!
+
+In 2021, we mostly write servers. That's why it's important to have a toolkit for working with 
+network sockets!
+
+In this post, we're going to do a quick tour of the tools we (Tilt-team) work with everyday,
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:22]:
```suggestion
In this post, we're going to do a quick tour of the tools we (the Tilt team) work with everyday,
```
Can we have something here to answer the question, "how will I, the reader, become more awesome by reading this post?" It's not clear to me what people are expected to gain here.

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+> `socat` stands for SOcket CAT. It is a utility for data transfer between two addresses.
+
+> I personally use `socat` not so much as a production debugging tool than for
+> troubleshooting local development issues (especially when developing network
+> services).
+
+Go ahead and [read it](https://copyconstruct.medium.com/socat-29453e9fc8a6) and come back. I'll wait.
+
+.
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:39]: Realistically, no one's gonna go read it right away. People will open a new tab that'll sit there for later. We can omit this multiline "ellipsis."

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+Can we do better?
+
+## Inside the Box: A port-forward as Just Another Server
+
+One way to think about `kubectl port-forward` is just another server. It's a very simple server!
+But it tastes great with lots of other tooling that we use to manage multiple servers and auto-restart
+them when they're not healthy.
+
+For example, I might use `entr` to restart the port-forward when I hit spacebar.
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:146]: I don't understand how spacebar does anything here? 🤔

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+new pod).
+
+Meanwhile, the borked port-forward is using up a terminal.
+
+Can we do better?
+
+## Inside the Box: A port-forward as Just Another Server
+
+One way to think about `kubectl port-forward` is just another server. It's a very simple server!
+But it tastes great with lots of other tooling that we use to manage multiple servers and auto-restart
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:143]:
```suggestion
But it goes great with lots of other tooling that we use to manage multiple servers and auto-restart
```
"tastes" in the context of kubectl port-fotward was a very weird synesthesia trip for a second there 😂

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+3) The fundamental Kubernetes pattern -- watch for changes to an API, and update
+our servers in response -- is really useful when managing servers!
+
+We're not building this API only for internal Tilt use. Tilt exposes them
+publicly, to help connect new tools that build on top of them.
+
+What will those new tools be? This post is already a bit long so it will have to
+wait for another post. But let us know if you'd like to see some ideas, or
+want to build on this API together :).
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:289]:
```suggestion
want to build on this API together. 😀
```

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    ...
+    Name:                    example-go-5cc87c754d-9ngn4
+    ...
+```
+
+The KubernetesDiscovery object contains a port-forward template. As pods are
+created, updated, and deleted, this object will select the "best" pod for
+portforwarding, and sets up a PortForward object to manage the connection.
+
+## Why am I telling you this???
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:269]: I think this section is good from an overview/philosophical perspective but going back to my first comment up top, what do readers gain from this exactly in practical terms?

"It's 2AM, I'm tired, and not very smart, connect the dots for me, what should I use this knowledge for?"

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+```python
+local_resource(
+  name='example-go-8000',
+  serve_cmd='kubectl port-forward deployment/example-go 8080:8000',
+  links=['http://localhost:8080'],
+  readiness_probe=probe(http_get=http_get_action(port=8080)))
+```
+
+Tilt will periodically check that the portforward is running. Other tools for
+managing multiple services (e.g., tmux) will work well too!
```
Comment (milas) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:175]:
```suggestion
managing multiple services (e.g. tmux) will work well too!
```
Errant comma

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 4 above)
Comment (milas) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md:143]: Inquiring minds want to know what `kubectl` tastes like and what type of wine it should be paired with

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 1 above)
Comment (milas) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md:22]: Yeah, I think there's a natural progression here of 1. `socat` is the bomb but it's the assembly of socket tools 2. `kubectl port-forward` is not ideal for part of daily workflow 3. Tilt is more ergonomic + more suited for dev loop

I like your line later:
> We're not building this API only for internal Tilt use. Tilt exposes them
> publicly, to help connect new tools that build on top of them.

I wonder if it makes sense to allude to that here somehow in the vein of "Tilt does this well AND with the new APIs you don't have to buy in to completely switching your local dev to Tilt to be able to benefit from it"

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 2 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: done

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 3 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: From the `entr` manual:

```
     entr listens for keyboard input and responds to the following commands:

     ⟨space⟩  Execute the utility immediately.  If the -r option is set this
              will terminate and restart the child process as if a file change
              event had occurred.
```

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 7 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: done

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 5 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: done

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 4 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: :wine_glass:

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 1 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: ya, i agree the intro is weak. i took some language from Milas' comment and added some mission statements. wdyt?

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 6 above)
Comment (nicks) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md:281]: ya, agreed, what do you think of the new intro?

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 3 above)
Comment (ellenkorbes) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: Oh, whoa. TIL.

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md (hunk 4 above)
Comment (ellenkorbes) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: It tastes like old beige keyboards.

Comment (ellenkorbes) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: Similar to worn money bills, but milder.

Comment (ellenkorbes) [reply blog/_posts/2021-07-08-portforwarding-is-awesome.md]: Cabernet Sauvignon?

File: blog/_posts/2021-07-08-portforwarding-is-awesome.md [hunk 8]
```diff
@@ -19,8 +19,23 @@ tools like `cat` for working with files!
...
+3. Tilt's `PortForward` API, a self-healing wrapper around `kubectl port-forward`.
+
+You may use a different set of socket tools! That's OK. 
+
+The goal of this post isn't to try to convince you that these particular tools
+are the best ones. Instead, I've been been thinking through what different tools
+are good for, look at how they interoperate with each other, and come up with a
+framework for figuring out which one is right for what I'm working on.
+
+But if you learn about a new tool in this post, that's fine too!
```
Comment (ellenkorbes) [inline blog/_posts/2021-07-08-portforwarding-is-awesome.md:39]: Yeah, this is great. But we don't have to be so apologetic—how about rephrasing this:

> You may use a different set of socket tools! That's OK. 
> The goal of this post isn't to try to convince you that these particular tools are the best ones. Instead, I've been been thinking through what different tools are good for, look at how they interoperate with each other, and come up with a framework for figuring out which one is right for what I'm working on.
> But if you learn about a new tool in this post, that's fine too!

As:

> I've been thinking through what different tools are good for, looking at how they interoperate with each other, and came up with a framework for figuring out which one is right for what I'm working on. 
> You don't have to use these tools in particular, but I hope you'll learn something new!

**Review by ellenkorbes — APPROVED** [review]

**Comment by ellenkorbes** [comment]
👍

**Pull request merged**